cpp/src/arrow/compute/exec/hash_join_node.cc

```diff
@@ -454,6 +456,172 @@ Status HashJoinSchema::CollectFilterColumns(std::vector<FieldRef>& left_filter,
+      ARROW_ASSIGN_OR_RAISE(ExecBatch key_batch, ExecBatch::Make(std::move(keys)));
+      RETURN_NOT_OK(Hashing32::HashBatch(key_batch, hashes.data(),
+                                         ctx_->cpu_info()->hardware_flags(), stack, 0,
+                                         key_batch.length));
+
+      eval_.received_filters_[ifilter]->Find(ctx_->cpu_info()->hardware_flags(),
+                                             key_batch.length, hashes.data(), bv.data());
+      arrow::internal::BitmapAnd(bv.data(), 0, selected.data(), 0, key_batch.length, 0,
+                                 selected.data());
+    }
+    auto selected_buffer =
+        arrow::internal::make_unique<Buffer>(selected.data(), bit_vector_bytes);
+    ArrayData selected_arraydata(boolean(), batch.length,
+                                 {nullptr, std::move(selected_buffer)});
+    Datum selected_datum(selected_arraydata);
+    FilterOptions options;
+    size_t first_nonscalar = batch.values.size();
+    for (size_t i = 0; i < batch.values.size(); i++) {
+      if (!batch.values[i].is_scalar()) {
+        ARROW_ASSIGN_OR_RAISE(batch.values[i],
+                              Filter(batch.values[i], selected_datum, options, ctx_));
+        first_nonscalar = std::min(first_nonscalar, i);
+        ARROW_DCHECK_EQ(batch.values[i].length(), batch.values[first_nonscalar].length());
+      }
+    }
+    // If they're all Scalar, then the length of the batch is the number of set bits
+    if (first_nonscalar == batch.values.size())
+      batch.length = arrow::internal::CountSetBits(selected.data(), 0, batch.length);
+    else
+      batch.length = batch.values[first_nonscalar].length();
+    return Status::OK();
+  }
+
+  Status BuildBloomFilter_exec_task(size_t thread_index, int64_t task_id);
+
+  Status BuildBloomFilter_on_finished(size_t thread_index) {
+    return build_.on_finished_(thread_index, std::move(build_.batches_));
+  }
+
+  // The Bloom filter is built on the build side of some upstream join. For a join to
+  // evaluate the Bloom filter on its input columns, it has to rearrange its input columns
+  // to match the column order of the Bloom filter.
+  //
+  // The first part of the pair is the HashJoin to actually perform the pushdown into.
+  // The second part is a mapping such that column_map[i] is the index of key i in
+  // the first part's input.
+  // If we should disable Bloom filter, returns nullptr and an empty vector, and sets
+  // the disable_bloom_filter_ flag.
+  std::pair<HashJoinNode*, std::vector<int>> GetPushdownTarget(HashJoinNode* start);
+
+  Result<util::TempVectorStack*> GetStack(size_t thread_index) {
+    if (!tld_[thread_index].is_init) {
+      RETURN_NOT_OK(tld_[thread_index].stack.Init(
+          ctx_->memory_pool(), 4 * util::MiniBatch::kMiniBatchLength * sizeof(uint32_t)));
+      tld_[thread_index].is_init = true;
+    }
+    return &tld_[thread_index].stack;
+  }
+
+  bool disable_bloom_filter_;
+  HashJoinNode* owner_;
```

Comment:
It seems this is only used to access the schema projection maps.  Maybe just take a pointer to that instead?  Then it would be clearer why this is needed.

Reply:
I also use `owner` to figure out the start point of the pushdown search and also for the ExecContext. That said, that's only needed inside of `Init`, so I've changed the member to just be a `HashJoinSchema`, but `Init` will still need the owner.